internal/node/syncer.go

```diff
@@ -17,32 +18,37 @@ type Syncer interface {
 	Addr() string
 	BroadcastHeader(h gateway.BlockHeader)
 	BroadcastTransactionSet([]types.Transaction)
-	Close() error
+	Close(ctx context.Context) error
 	Connect(ctx context.Context, addr string) (*syncer.Peer, error)
 	Peers() []*syncer.Peer
 }
 
-type nodeSyncer struct {
+type closeSyncer struct {
 	*syncer.Syncer
-	l net.Listener
+	l       net.Listener
+	errChan chan error
 }
 
-func (s *nodeSyncer) Close() error {
-	return s.l.Close()
+func (s *closeSyncer) Close(ctx context.Context) error {
+	if err := s.l.Close(); err != nil {
```

Comment:
Are you sure this fixes the issue? Looking at the `Syncer` it doesn't seem like waiting for `Run` to return is necessarily a guarantee for all goroutines (e.g. the one in `runPeer`) to be done.

I think the proper way would be an actual `Close` method on the `Syncer` with a `WaitGroup`.

Reply:
I think I proposed that once but I'm not sure how that ended. Hm well no I'm not sure if it fixes the issue because it's hard to reproduce it, I did restart a running node a couple of times to verify but that's not much of a test. The `Run` docs on the `syncer` say to close the listener so I made the assumption that was how it was intended to be used. I can bring it up with Luke, adding a `Close` on the `syncer` would definitely be more ideal, then we don't need the weird wrapper at all.

```go
// Run spawns goroutines for accepting inbound connections, forming outbound
// connections, and syncing the blockchain from active peers. It blocks until an
// error occurs, upon which all connections are closed and goroutines are
// terminated. To gracefully shutdown a Syncer, close its net.Listener.
func (s *Syncer) Run() error {
```

Reply:
@lukechampine could you take a look?

Reply:
Ah I was about to create an issue on the repo but this works too 😁

Reply:
I can push a fix for `Run` returning before all the goroutines have exited. And I think I'm ok adding a `Close` method; it's definitely annoying that you have to store the `net.Listener` in order to shutdown.